rmf_task/include/rmf_task/Request.hpp

```diff
@@ -34,25 +35,45 @@ class Request
 {
 public:
 
-  using SharedPtr = std::shared_ptr<Request>;
+  /// Constructor
+  ///
+  /// \param[in] id
+  ///   The unique id for this task
+  ///
+  /// \param[in] earliest_start_time
+  ///   The earliest time this task should begin execution. This is usually the
+  ///   requested start time for the task.
+  ///
+  /// \param[in] priority
+  ///   The priority for this task. This is provided by the CostCalculator. For
+  ///   requests that do not have any priority this is a nullptr
+  Request(
+    std::string& id,
+    rmf_traffic::Time earliest_start_time,
+    ConstPriorityPtr priority);
 
   /// Get the id of the task
-  virtual std::string id() const = 0;
+  std::string id() const;
+
+  /// Get the earliest start time that this task may begin
+  rmf_traffic::Time earliest_start_time() const;
+
+  /// Get the priority scheme of this task
+  ConstPriorityPtr priority() const;
 
   /// Estimate the state of the robot when the task is finished along with the
   /// time the robot has to wait before commencing the task
-  virtual rmf_utils::optional<Estimate> estimate_finish(
+  virtual std::optional<Estimate> estimate_finish(
```

Comment:
We should never ever ever mix together concrete types and abstract interfaces.

Let's define a new class to be the abstract interface for these functions, and then add it as a field to the `Request` class. For example:

```
class Request
{
public:
  class Description
  {
  public:
    virtual std::optional<Estimate> estimate_finish(...) const = 0;
    virtual rmf_traffic::Duration invariant_duration() const = 0;
    virtual ~Type() = default;
  };
  using DescriptionPtr = std::shared_ptr<const Description>;

  Request(
    const std::string& id,
    rmf_traffic::Time earliest_start_time,
    ConstPriorityPtr priority,
    DescriptionPtr description);
};
```

Reply:
I'm having trouble refactoring as suggested as the `estimate_finish()` function internally requires the `estimate_start_time` field. Would it be okay to move `earliest_start_time()` and `id()` into `Description` as virtual functions? I would like to move `id()` as well since this field is sometimes populated internally as in the case of automatically added tasks such as `ChargeBattery`. 
Proposed design:
```
class Request
{
public:
  class Description
  {
  public:

    /// The unique id for this request
    virtual const std::string& id() const = 0;

    /// Get the earliest time that this request may begin
    rmf_traffic::Time earliest_start_time() const = 0;

    /// Estimate the state of the robot when the request is finished along with the
    /// time the robot has to wait before commencing the request
    virtual std::optional<Estimate> estimate_finish(
      const agv::State& initial_state,
      const agv::Constraints& task_planning_constraints,
      const std::shared_ptr<EstimateCache> estimate_cache) const = 0;
    
    /// Estimate the invariant component of the request's duration
    virtual rmf_traffic::Duration invariant_duration() const = 0;

    virtual ~Description() = default;
  };
  using DescriptionPtr = std::shared_ptr<Description>;

  /// Constructor
  ///
  /// \param[in] earliest_start_time
  ///   The earliest time this request should begin execution. This is usually the
  ///   requested start time for the request.
  ///
  /// \param[in] priority
  ///   The priority for this request. This is provided by the Priority Scheme. For
  ///   requests that do not have any priority this is a nullptr.
  ///
  /// \param[in] description
  ///   The description for this request
  Request(
    ConstPriorityPtr priority,
    DescriptionPtr description);

  /// Get the priority of this request
  ConstPriorityPtr priority() const;

  /// Get the description of this request
  const DescriptionPtr& description() const;

  class Implementation;
private:
  rmf_utils::impl_ptr<Implementation> _pimpl;
};
```

Reply:
I think the abstraction that I proposed is the right way to do it.

I can think of two ways to deal with the earliest start time:

1. Design the factory functions for the various `Request::Description` implementations to take in an `earliest_start_time` argument, and have the `Request::Description` internally save its own copy for that field. We'd be passing the `earliest_start_time` as an argument to both the constructor of `Request` and the factory functions of `Request::Description`, so there's some duplication of truth, but I think that's acceptable in this case.
2. Have the `estimate_finish` function accept an `earliest_start_time` argument. This may be preferable if the `Request::Description` objects don't need to assume that the `earliest_start_time` will always be the same value, but it may be problematic if we ever want to allow them to pre-compute and/or cache anything based on the earliest start time of the task. They should still be able to do caching by maintaining a `std::unordered_map<rmf_traffic::Time, ...>` cache, it'll just be less convenient.

> I would like to move id() as well since this field is sometimes populated internally as in the case of automatically added tasks such as ChargeBattery.

I don't see any reason the ID can't be automatically generated for the `Request` internally when automatically generating a battery charging task. You would just have a battery-specific factory function that returns a `Request` object, and that factory function can generate the ID and use it to construct the `Request` object before returning that `Request` to you.

Overall I don't think there's a compelling reason to have those fields be virtual, since I can't conceive of a situation where their behavior would be implementation dependent. The start time and ID fields should both be values that get set during `Request` construction, and their accessor functions should just return whatever that original value was.

Reply:
Thanks for the detailed feedback. I will go with the first suggestion and store a copy of the `earliest_start_time` inside `Request::Description`. I didn't consider having two factory functions, one for generating a request-specific `Request` objects and the other for the request-specific `Request::Description` objects. I see how its possible to keep these fields concrete.

Reply:
I've updated the APIs accordingly
https://github.com/osrf/rmf_core/pull/278/commits/55d01babe67efd82f48ebf5dd1f684c6ca187b88
https://github.com/osrf/rmf_core/pull/278/commits/ed42b6f35c46dbce3d52ede0748bb762b076a4c8
https://github.com/osrf/rmf_core/pull/278/commits/e0a7c852e0a822b7bad4b913e236b7b2733e339b